Objects/codeobject.c

```diff
@@ -420,6 +421,10 @@ _PyCode_New(struct _PyCodeConstructor *con)
         PyErr_NoMemory();
         return NULL;
     }
+    co->co_filename = NULL;
+    co->co_name = NULL;
+    co->co_qualname = NULL;
+
```

Comment:
Hm, I changed my mind again. I think we should leave init_code() unchanged (storing values without looking) and in `_PyCode_Update()` we explicitly DECREF co_name, co_qualname and co_filename (because on entering there we know they are valid).

Moreover, in `_PyCode_Update()` we should actually preserve the original co_filename, and if it's different, overwrite the co_filename of any (newly created) dehydrated code objects, probably using the same logic as update_compiled_module() in import.c. In case you wonder why, the reason is to preserve the filename set at the toplevel by `_imp.fix_co_filename()` under some circumstances (called from `_bootstrap_external.py`) -- there's an obscure test in Lib/test/test_import/__init__.py (test_incorrect_code_name) that comes down to verifying that if a module is loaded from bytecode, the co_filename of all the loaded code objects is where it's loaded from, not whatever was marshaled into the file. The existing logic doesn't set this for code objects that are skipped by lazy loading (I already had to fix a crash there due to co_consts being NULL), and hydrating currently overwrites the fixed co_filename with the incorrect one read from the marshal data. (Arguably we should not serialize co_filename at all, and only set it when loading. But that's probably going to break some other test...)

Reply:
Yes, I saw that test but didn't understand it. I'll try this in a new PR. I think update_code_filenames needs to move from import.c to codeobject.c.

Reply:
I spent some quality time with both C and Python debuggers before I understood the feature that it tests. :/ The key is the _imp.fix… call in _bootstrap_external.py.

Reply:
See https://github.com/iritkatriel/cpython/pull/21